ostd/src/arch/riscv/trap/mod.rs

```diff
@@ -0,0 +1,43 @@
+// SPDX-License-Identifier: MPL-2.0
+
+//! Handles trap.
+
+mod trap;
+
+pub use trap::{GeneralRegs, TrapFrame, UserContext};
+
+use crate::cpu_local_cell;
+
+cpu_local_cell! {
+    static IS_KERNEL_INTERRUPTED: bool = false;
+}
+
+/// Initialize interrupt handling on RISC-V.
+pub unsafe fn init(on_bsp: bool) {
+    self::trap::init();
+}
+
+/// Returns true if this function is called within the context of an IRQ handler
+/// and the IRQ occurs while the CPU is executing in the kernel mode.
+/// Otherwise, it returns false.
+pub fn is_kernel_interrupted() -> bool {
+    IS_KERNEL_INTERRUPTED.load()
+}
+
+/// Handle traps (only from kernel).
+#[no_mangle]
+extern "C" fn trap_handler(f: &mut TrapFrame) {
+    use riscv::register::scause::Trap;
+
+    IS_KERNEL_INTERRUPTED.store(true);
+    match riscv::register::scause::read().cause() {
+        Trap::Interrupt(_) => todo!(),
+        Trap::Exception(e) => {
+            let stval = riscv::register::stval::read();
+            panic!(
+                "Cannot handle kernel cpu exception: {e:?}. stval: {stval:#x}, trapframe: {f:#x?}.",
+            );
+        }
+    }
+    IS_KERNEL_INTERRUPTED.store(false);
```

Comment:
I believe that IS_KERNEL_INTERRUPTED should only be set in the `Trap::Interrupt(_) => ` branch. If the calling code is in a CPU exception handler, `is_kernel_interrupted()` just returns false.